crates/js-component-bindgen/src/intrinsics.rs

```diff
@@ -114,6 +119,117 @@ pub fn render_intrinsics(
+                    if (realloc) {
+                        address = realloc(0, 0, 4, 1024);
+                        new Int32Array(memory.buffer, address).set([address + 8, address + 1024]);
+                    } else {
+                        address = 16;
+                        new Int32Array(memory.buffer, address).set([address + 8, address + 1024]);
+                    }
+                    asyncifyModules.push({ instance, memory, address });
+                  }
+                  return instance;
+                }
+                function asyncifyState() {
+                  return asyncifyModules[0]?.instance.exports.asyncify_get_state();
+                }
+                function asyncifyAssertNoneState() {
+                  let state = asyncifyState();
+                  if (state !== 0) {
+                    throw new Error(`reentrancy not supported, expected asyncify state '0' but found '${state}'`);
+                  }
+                }
+            "),
+
+            Intrinsic::AsyncifySyncInstantiate => output.push_str("
+                const asyncifyModules = [];
+                let asyncifyPromise;
+                let asyncifyResolved;
+                function asyncifyInstantiate(module, imports) {
+                  const instance = instantiateCore(module, imports);
+                  const memory = instance.exports.memory || (imports && imports.env && imports.env.memory);
+                  const realloc = instance.exports.cabi_realloc || instance.exports.cabi_export_realloc;
+                  if (instance.exports.asyncify_get_state && memory) {
+                    let address;
+                    if (realloc) {
+                        address = realloc(0, 0, 4, 1024);
+                        new Int32Array(memory.buffer, address).set([address + 8, address + 1024]);
+                    } else {
+                        address = 16;
+                        new Int32Array(memory.buffer, address).set([address + 8, address + 1024]);
+                    }
+                    asyncifyModules.push({ instance, memory, address });
+                  }
+                  return instance;
+                }
+                function asyncifyState() {
+                  return asyncifyModules[0]?.instance.exports.asyncify_get_state();
+                }
+                function asyncifyAssertNoneState() {
+                  let state = asyncifyState();
+                  if (state !== 0) {
+                    throw new Error(`reentrancy not supported, expected asyncify state '0' but found '${state}'`);
+                  }
+                }
+            "),
+
+            Intrinsic::AsyncifyWrapExport => output.push_str("
+                function asyncifyWrapExport(fn) {
+                  return async (...args) => {
+                    if (asyncifyModules.length === 0) {
+                        throw new Error(`none of the Wasm modules were processed with wasm-opt asyncify`);
```

Comment:
If we didn't have the fall-through path on the asyncify instantiate, would we be able to turn this into a static instead of a runtime error?